frontend/src/metabase/components/form/FormField.jsx

```diff
@@ -1,100 +1,102 @@
-/* eslint-disable react/prop-types */
-import React, { Component } from "react";
+import React from "react";
 import PropTypes from "prop-types";
 import cx from "classnames";
 
 import Tooltip from "metabase/components/Tooltip";
 
-import { FieldRow, Label, InfoIcon } from "./FormField.styled";
+import { FieldRow, Label, InfoIcon, InputContainer } from "./FormField.styled";
 
-export default class FormField extends Component {
-  static propTypes = {
-    field: PropTypes.object,
-    formField: PropTypes.object,
+const formFieldCommon = {
+  title: PropTypes.string,
+  description: PropTypes.string,
+  info: PropTypes.string,
+  hidden: PropTypes.bool,
+  horizontal: PropTypes.bool,
+};
 
-    // redux-form compatible:
-    name: PropTypes.string,
-    error: PropTypes.any,
-    visited: PropTypes.bool,
-    active: PropTypes.bool,
+const propTypes = {
+  ...formFieldCommon,
 
-    hidden: PropTypes.bool,
-    title: PropTypes.string,
-    description: PropTypes.string,
-    info: PropTypes.string,
+  field: PropTypes.object,
+  formField: PropTypes.shape({
+    ...formFieldCommon,
+    type: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
+  }),
 
-    children: PropTypes.oneOfType([
-      PropTypes.arrayOf(PropTypes.node),
-      PropTypes.node,
-    ]),
+  // redux-form compatible:
+  name: PropTypes.string,
+  error: PropTypes.any,
+  visited: PropTypes.bool,
+  active: PropTypes.bool,
+
+  children: PropTypes.oneOfType([
+    PropTypes.arrayOf(PropTypes.node),
+    PropTypes.node,
+  ]),
+  className: PropTypes.string,
+};
+
+const ALL_DOT_CHARS = /\./g;
+
+function FormField(props) {
+  const {
+    className,
+    formField,
+    title = formField && formField.title,
+    description = formField && formField.description,
+    info = formField && formField.info,
+    hidden = formField && (formField.hidden || formField.type === "hidden"),
+    horizontal = formField &&
+      (formField.horizontal || formField.type === "boolean"),
+    children,
+  } = props;
```

Comment:
Forgive my ignorance, but  `title`, `description`, `info`, `hidden` and `horizontal` are kind of artificial props, is this the case?

If so, this would be a stylistic nitpick, but wdyt of writing this kind of like

```js
function FormField({children, className, formField}) {
  const { description, info, title } = formField || {}
  const hidden = formField && (formField.hidden || formField.type === "hidden")
  const horizontal = formField && (formField.horizontal || formField.type === "boolean")
  …
    
```
?

Reply:
can't wait for optional chaining in this codebase too…

Reply:
By default, the component uses `title`, 'description, etc. from `formField`, but they also can be passed directly to overwrite the `formField` value. Example:

```
<FormField formField={{ title: "A" }} /> // title will be A
<FormField title="B" formField={{ title: "A" }} /> // title will be B
```

Your refactoring looks much better to me, I'll update the PR. Thank you!